File: batou-rules/testdata/fixtures/java/safe/ZipSlipSafe.java
import java.io.*;
import java.util.zip.*;

public class ZipSlipSafe {
    public void extractZip(String zipFile, String destDir) throws Exception {
        File destDirectory = new File(destDir);
        ZipInputStream zis = new ZipInputStream(new FileInputStream(zipFile));
        ZipEntry entry;
        while ((entry = zis.getNextEntry()) != null) {
            File destFile = new File(destDir, entry.getName());
            // SAFE: validate canonical path is within destination directory
            String canonicalPath = destFile.getCanonicalPath();
            String canonicalDest = destDirectory.getCanonicalPath() + File.separator;
            if (!canonicalPath.startsWith(canonicalDest)) {
                throw new SecurityException("Entry outside target dir: " + entry.getName());
            }
            destFile.getParentFile().mkdirs();

            FileOutputStream fos = new FileOutputStream(destFile);
            byte[] buffer = new byte[1024];
            int len;
            while ((len = zis.read(buffer)) > 0) {
                fos.write(buffer, 0, len);
            }
            fos.close();
        }
        zis.close();
    }
}

File: batou-rules/testdata/fixtures/javascript/safe/zip_slip_safe.ts
import * as fs from 'fs';
import * as path from 'path';
import * as yauzl from 'yauzl';

function extractZipSafe(zipPath: string, destDir: string) {
  yauzl.open(zipPath, { lazyEntries: true }, (err, zipfile) => {
    zipfile.readEntry();
    zipfile.on('entry', (entry) => {
      const destPath = path.resolve(destDir, entry.fileName);
      // SAFE: validate resolved path starts with destination directory
      if (!destPath.startsWith(path.resolve(destDir) + path.sep)) {
        throw new Error(`Zip slip detected: ${entry.fileName}`);
      }
      const writeStream = fs.createWriteStream(destPath);
      zipfile.openReadStream(entry, (err, readStream) => {
        readStream.pipe(writeStream);
      });
      zipfile.readEntry();
    });
  });
}
